Implement digamma

zou3519 opened the pull request:
Implements `torch.digamma` and `tensor.digamma()` as per #678

cc @fritzo 

### Test Plan
New unit tests. cpu unit tests test some edge cases, gpu unit test compares output of cpu to gpu

zou3519 committed:
Implement digamma

fritzo commented on test/test_torch.py:
```diff
@@ -194,6 +194,24 @@ def test_sinh(self):
 
+    def test_digamma(self):
+        y = torch.Tensor([-10, 0])
+        x = torch.Tensor([-0.1, 3, 999])
```
You could also check against hard-coded values generated by `scipy.special.digamma()`, e.g.:
```py
>>> import numpy as np
>>> import scipy.special
>>> x = 10.0 ** np.arange(-5, 5, 0.5)
>>> x
array([  1.00000000e-05,   3.16227766e-05,   1.00000000e-04,
         3.16227766e-04,   1.00000000e-03,   3.16227766e-03,
         1.00000000e-02,   3.16227766e-02,   1.00000000e-01,
         3.16227766e-01,   1.00000000e+00,   3.16227766e+00,
         1.00000000e+01,   3.16227766e+01,   1.00000000e+02,
         3.16227766e+02,   1.00000000e+03,   3.16227766e+03,
         1.00000000e+04,   3.16227766e+04])
>>> scipy.special.digamma(x)
array([ -1.00000577e+05,  -3.16233538e+04,  -1.00005771e+04,
        -3.16285436e+03,  -1.00057557e+03,  -3.16799792e+02,
        -1.00560885e+02,  -3.21491437e+01,  -1.04237549e+01,
        -3.31321826e+00,  -5.77215665e-01,   9.84925052e-01,
         2.25175259e+00,   3.43798293e+00,   4.60016185e+00,
         5.75488076e+00,   6.90725520e+00,   8.05888970e+00,
         9.21029037e+00,   1.03616171e+01])
```

fritzo reviewed (commented):
Thanks for implementing this!

gchanan commented:
fwiw I'm adding the ability to write efficient pointwise native functions in ATen directly.

fritzo commented:
Does this PR also register `digamma` as the derivative of `lgamma`? (I'm not sure how that works.)

fritzo commented on tools/autograd/derivatives.yaml:
```diff
@@ -305,7 +305,10 @@
 
 - name: lgamma(Tensor self)
-  self: not_implemented("lgamma")
+  self: grad * digamma(self)
```
Do I understand correctly that this is all that is needed to register `digamma` as the grad of `lgamma`?

apaszke replied:
Yes, that's correct

apaszke commented on aten/src/TH/generic/THTensorMath.c:
```diff
@@ -3092,6 +3092,81 @@ void THTensor_(atan2)(THTensor *r_, THTensor *tx, THTensor *ty)
+    return THTensor_(digamma_one)(1 - x) - PI / tan(PI * x);
+  }
+
+	// Push x to be >= 10
```
nit: tabs

apaszke commented on aten/src/TH/generic/THTensorMath.c:
```diff
@@ -3092,6 +3092,81 @@ void THTensor_(atan2)(THTensor *r_, THTensor *tx, THTensor *ty)
+}
+
+/*
+ * Algorithm adapted from Cephes
```
Does adapted == pretty much copied, or did you rewrite it based on Cephes? I'm not sure what's the license, but we might need to include it here.

zou3519 replied:
I think the answer to that is pretty much copied: the algorithm is the same, except for one case (integers from 1 - 10). Here's the original cephes code I used for reference: https://github.com/jeremybarnes/cephes/blob/master/misc/psi.c -- I can include the license there here.

apaszke replied:
Cephes doesn't state its license clearly except from saying that it "can be used freely". I think it's best to ask the author, just to be sure. Aparently [that's what authors of torch-cephes did](https://github.com/deepmind/torch-cephes/blob/master/LICENSE.txt).

apaszke commented on test/test_cuda.py:
```diff
@@ -1032,6 +1033,23 @@ def test_var_stability(self):
         tensor = tensor.unsqueeze(1)
         self.assertEqual(tensor.var(0)[0], 0.03125)
 
+    def test_digamma(self):
```
Any reason why `digamma` can't be tested like all other functions?

zou3519 replied:
The values of digamma on gpu when compared to cpu are the same to 6 digits. I tried making this a normal test and some output values were high (~100), resulting in an error of around 1e-3, which is above the error threshold

apaszke reviewed (commented):
Looks good. We should resolve the licensing issues before merging this.

fritzo commented:
FYI this will merge conflict with #3978 where I've implemented a makeshift finite-difference version of `THTensor(digamma_one)`. If #3978 merges before this PR, you can simply delete my makeshift version.

fritzo commented:
@zou3519 can I help with this at all? Now that you've done the hard part, I'm happy to resolve merge conflicts and add tests. (We're looking forward to using this in Pyro, and we already have a wrapper to use `torch.distributions.Gamma`).

zou3519 commented:
@fritzo I haven't reached out to the author about the licensing yet. I'll do that tomorrow and we'll see how that goes :)

fritzo commented:
I see, thanks for letting me know!

ezyang commented:
Any update on the license status?

zou3519 commented:
Sent an email a few days ago to the author, haven't heard back yet

fritzo commented:
If we still haven't heard from the author, I could throw together a little PR that exposes the makeshift finite-difference implementation of digamma that we're already using internally in a few places. This would at least unblock users of Gamma, Beta, and Dirichlet distributions.
- https://github.com/pytorch/pytorch/blob/15b657af843d0dd95fa43d44348161d3785ed0e9/aten/src/ATen/native/Distributions.cpp#L21
- https://github.com/pytorch/pytorch/blob/15b657af843d0dd95fa43d44348161d3785ed0e9/aten/src/TH/generic/THTensorMath.c#L3483
- https://github.com/pytorch/pytorch/blob/15b657af843d0dd95fa43d44348161d3785ed0e9/torch/distributions/utils.py#L6

apaszke commented:
I think merging an approximation is good for now

fritzo commented:
Hmm I looked into it but it appears to be quite complex now that our functions using `digamma` are spread across C code in `aten/src/TH/generic/THTensorMath.c` and C++ code in `aten/src/ATen/native/Distributions.cpp`. @gchanan would it be difficult for you to move `dirichlet_grad()` to C++ as well so we can implement `digamma()` in a single place?

gchanan commented:
@fritzo sure, I'll take a look at.

zou3519 commented:
Licensing issues resolved, PR rebased. I was told that "You are welcome to modify and distribute the library under BSD license." so I've added the original copyright notices into the code as comments.

fritzo commented:
That's great news, @zou3519 !

zou3519 committed:
Better test coverage for digamma.

Added test to check digamma float vs double.
TestCuda.test_digamma checks the CUDA {float, double} implementation
against the CPU {float, double} implementation.

zou3519 committed:
Replace usages of finite-differences digamma with newly implemented digamma

wranai commented:
Any news on this? It seems only the GPU build failed on Windows last November and that only because it run out of memory, but everything else was okay.

fritzo commented:
@wranai A simpler implementation of digamma and trigamma is already in master. Try `torch.digamma`
